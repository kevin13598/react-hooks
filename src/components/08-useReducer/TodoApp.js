import React, { useReducer }  from 'react';
import { todoReducer } from './todoReducer';
import { useEffect } from 'react';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';

import './styles.css';

const init = () => {

    return JSON.parse(localStorage.getItem('todos')) || [];

}

export const TodoApp = () => {

    const [todos, dispatch] = useReducer(todoReducer, [], init);

    

    useEffect( () => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos]);

    const handleDelete = (id) => {
        
        dispatch({
            type: 'delete',
            payload: id
        });

    }

    const handleToggle = (id) => {
        dispatch({
            type: 'toggle',
            payload: id
        })
    }

    const handleAddTodo = (newTodo) => {

        dispatch({
            type: 'add',
            payload: newTodo
        });
    }

    return (
        <div>
            <h1>TodoApp ({todos.length})</h1>
            <hr/>

           <div className="row">

               <div className="col-7">

                    <TodoList 
                        todos={todos} 
                        onDelete={handleDelete} 
                        onToggle={handleToggle} 
                    />

               </div>

               <div className="col-5">
                    <TodoAdd handleAddtodo = {handleAddTodo} />
               </div>

           </div>

        </div>
    )
}
