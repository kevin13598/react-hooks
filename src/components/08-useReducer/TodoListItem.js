import React from 'react';

export const TodoListItem = ({todo, index, onToggle, onDelete}) => {

    return (
        <li
            key={todo.id}
            className="list-group-item"
        >
            <p 
                className={`${todo.done && 'complete'}`}
                onClick= {() => onToggle(todo.id)}
            >
                {index+1}. {todo.desc}
            </p>

            <button
                className="btn btn-danger"
                onClick = {() => onDelete(todo.id)}
            >
                Borrar
            </button>
    </li>                     
    )
}
