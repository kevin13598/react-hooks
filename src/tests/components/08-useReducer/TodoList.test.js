import React from 'react';
import { shallow } from 'enzyme';
import { demoTodos } from '../../fixtures/demoTodos';
import { TodoList } from '../../../components/08-useReducer/TodoList';

describe('Pruebas en <TodoList />', () => {

   const onDelete = jest.fn();
   const onToggle = jest.fn();

   
    const wrapper = shallow(
        <TodoList 
            todos = {demoTodos}
            onDelete = {onDelete}
            onToggle = {onToggle}
        />
    );

    test('debe de mostrarse correctamente', () => {

        expect(wrapper).toMatchSnapshot();

    });
    
    test('debe de tener dos <TodoListItem/>', () => {

        expect(wrapper.find('TodoListItem').length).toBe(demoTodos.length);

        expect(wrapper.find('TodoListItem').at(0).prop('onDelete')).toEqual(expect.any(Function));

    });
});