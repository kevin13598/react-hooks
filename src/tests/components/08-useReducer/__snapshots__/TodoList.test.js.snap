// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoList /> debe de mostrarse correctamente 1`] = `
<ul
  className="list-group list-group-flush"
>
  <TodoListItem
    index={0}
    key="1"
    onDelete={[MockFunction]}
    onToggle={[MockFunction]}
    todo={
      Object {
        "desc": "Aprender React",
        "done": false,
        "id": 1,
      }
    }
  />
  <TodoListItem
    index={1}
    key="2"
    onDelete={[MockFunction]}
    onToggle={[MockFunction]}
    todo={
      Object {
        "desc": "Aprender Mongo",
        "done": false,
        "id": 2,
      }
    }
  />
</ul>
`;
